tests/test-daemon.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
+# This test daemonizes an nmsgtool and checks its output to make sure it is
+# equivalent to the input file.  It also tests the PID functionality present
+# in daemon.c.
+
+script_dir=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+nmsgtool_test=$script_dir/../src/nmsgtool
+infile=$script_dir/private-tests/202_sample.nmsg
+outfile=/tmp/testdaemon.nmsg
+pidfile=/tmp/testdaemon.pid
+retval=0
+
+check() {
+	if [ $? = "0" ]; then
+		echo "PASS: $*"
+	else
+		echo "FAIL: $*"
+		retval=1
+	fi
+}
+
+killproc () {
+	kill $1 >/dev/null 2>&1
+	killprocess_pid=$!
+	wait $killprocess_pid
+}
+
+echo "Testing nmsgtool daemonization: "
+
+# Start a daemonized nmsgtool and wait for it to finish.
+$nmsgtool_test -ddddd -D -r $infile -w $outfile &
+sleep 1
+killproc $!
```

Comment:
Note that this kill doesn't really do anything. If there were no `-D` parameter then the special value `$!` would correctly point to what you want it to, but the first step of daemonization is to `fork()`, allowing the parent process to exit while continuing to execute in the child. So here the return value of `$!` will point to the stale parent pid that has already exited, and not the new process that becomes the running nmsgtool instance that you're interested in.

Reply:
Fixed